File: ui/src/app/model/product.ts
import { MulticastGroup } from "./multicast-group";



export class Product{

    id: number;
    name: string;
    multicastGroups: MulticastGroup[]
  
    constructor(id: number = null, name: string = null,
      multicastGroups: MulticastGroup[] = []) {
      this.id = id;
      this.name = name;
      this.multicastGroups = multicastGroups;
      
    }
    
    getId(): number {
      return this.id;
    }
    getName(): string {
      return this.name;
    }
    getMulticastGroups(): MulticastGroup[] {
      return this.multicastGroups;
    }
    setMulticastGroups(multicastGroups: MulticastGroup[],
      sort: (a: MulticastGroup, b: MulticastGroup) => number = 
      (a: MulticastGroup, b: MulticastGroup) => {
        if(a.getName() > b.getName()) return 1;
        if(a.getName() == b.getName()) return 0;
        return -1;
      }) {
      this.multicastGroups = multicastGroups.sort(sort);
    }

    equals(other: Product): boolean {
      return this.getName() === other.getName();
    }
  }
  

File: src/main/java/com/mni/model/product/Product.java
package com.mni.model.product;


import com.mni.model.multicastgroup.MulticastGroup;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.Collection;
import java.util.stream.Collectors;

@Entity
public class Product {

    @Id
    @GeneratedValue
    private Long id;

    @Column(unique=true)
    private String name;

    @NotNull
    @ManyToMany
    Collection<MulticastGroup> multicastGroups = new ArrayList();

    public Product(){}

    public Product(Long id, String name, Collection<MulticastGroup> multicastGroups) {
        this.id = id;
        this.name = name;
        this.multicastGroups = multicastGroups;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Collection<MulticastGroup> getMulticastGroups() { return multicastGroups; }

    public void setMulticastGroups(Collection<MulticastGroup> multicastGroups) {
        if(this.multicastGroups == null) this.multicastGroups = new ArrayList();
        this.multicastGroups.clear();
        this.multicastGroups.addAll(multicastGroups);
    }

    public void addMulticastGroup(MulticastGroup multicastGroup) {
        if(this.multicastGroups == null) this.multicastGroups = new ArrayList();
        this.multicastGroups.add(multicastGroup);
    }

    public boolean removeMulticastGroup(MulticastGroup multicastGroup) {
        if(this.multicastGroups == null) this.multicastGroups = new ArrayList();
        return this.multicastGroups.removeIf(group -> group.getId() == multicastGroup.getId());
    }

    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", multicastGroups ids=" + multicastGroups.stream().map(g -> g.getId()).collect(Collectors.toList()) +
                '}';
    }
}
